### PR TITLE
Bonnie v2.1 vsac refactor

### PR DESCRIPTION
**Pull request by hossenlopp**
Creation of a new VSAC API class in bonnie_bundler. Now supporting the release parameter and utility functions of the API. Some reworking of parameter passing occurred in value set loading classes. This will be expanded on in task to support release parameter.

Pull requests into Bonnie Bundler require the following. Submitter and reviewer should :white_check_mark: when done. For items that are not-applicable, note it's not-applicable ("N/A") and :white_check_mark:.

**Submitter:**
- [x] This pull request describes why these changes were made.
- [x] This PR is into the correct branch.
- [x] JIRA ticket for this PR: https://jira.mitre.org/browse/BONNIE-1338
- [x] JIRA ticket links to this PR
- [x] Code diff has been done and been reviewed (it **does not** contain: additional white space, not applicable code changes, debug statements, etc.)
- [x] Tests are included and test edge cases
- [x] Tests have been run locally and pass (remember to update Gemfile when applicable)
- [x] Code coverage has not gone down and all code touched or added is covered.
    * In rare situations, this may not be possible or applicable to a PR. In those situations:
        1. Note why this could not be done or is not applicable here:
        2. Add TODOs in the code noting that it requires a test
        3. Add a JIRA task to add the test and link it here:
    * Base: 694 / 1162 LOC (59.72%)
    * PR: 801 / 1272 LOC (62.97%)
- [x] Automated regression test(s) pass N/A Does not affect calculation

If JIRA tests were used to supplement or replace automated tests: N/A
- [x] JIRA test links: N/A
- [x] Justification for using JIRA tests: N/A
- [x] JIRA tests have been added to sprint N/A

**Reviewer 1:**

Name:
- [x] Code is maintainable and reusable, reuses existing code and infrastructure where appropriate, and accomplishes the task’s purpose
- [x] The tests appropriately test the new code, including edge cases
- [x] You have tried to break the code

If JIRA tests were used to supplement or replace automated tests: N/A
- [x] JIRA tests have been run and pass N/A
- [x] You agree with the justification for use of JIRA tests or have provided input on why you disagree N/A

**Reviewer 2:**

Name: @losborne
- [x] Code is maintainable and reusable, reuses existing code and infrastructure where appropriate, and accomplishes the task’s purpose
- [x] The tests appropriately test the new code, including edge cases
- [x] You have tried to break the code

If JIRA tests were used to supplement or replace automated tests: N/A
- [x] JIRA tests have been run and pass N/A
- [x] You agree with the justification for use of JIRA tests or have provided input on why you disagree N/A

### CODE REVIEW

File: lib/util/vsac_api.rb [hunk 1]
```diff
@@ -0,0 +1,245 @@
...
+        super('VSAC ULMS credentials were not provided.')
+      end
+    end
+
+    # Raised when the arguments passed in are bad.
+    class VSACArgumentError < VSACError
+    end
+
+    class VSACAPI
+      DEFAULT_PROGRAM = "CMS eCQM"
```
Comment (edeyoung) [inline lib/util/vsac_api.rb]: move to config file?

File: lib/util/vsac_api.rb [hunk 2]
```diff
@@ -0,0 +1,245 @@
...
+            @config = symbolized_config
+          else
+            raise VSACArgumentError.new("Required param :config is missing required URLs.")
+          end
+        end
+
+        # if a ticket_granting_ticket was passed in, check it and raise errors if found
+        # username and password will be ignored
+        if !options[:ticket_granting_ticket].nil?
+          tgt = options[:ticket_granting_ticket]
```
Comment (edeyoung) [inline lib/util/vsac_api.rb]: does `tgt` stand for ticket granting ticket? If so, also looks like target, so consider spelling out more.

File: lib/util/vsac_api.rb (hunk 1 above)
Comment (hossenlopp) [reply lib/util/vsac_api.rb]: This is a fail safe program to use for if the config doesn't supply a program. If no profile is provided to `get_releases_for_program`, it checks the config first, then will fallback to this constant.

Should I remove this constant and require the config to have a program?

File: lib/util/vsac_api.rb (hunk 2 above)
Comment (hossenlopp) [reply lib/util/vsac_api.rb]: I'll spell this out.

File: lib/util/vsac_api.rb (hunk 1 above)
Comment (edeyoung) [reply lib/util/vsac_api.rb]: what is done with the other vsac parameters? if they're required in teh config, i'd require this as well - i think this one is as important. 

what happens if someone has a default release in an different program than the default program?

File: lib/util/vsac_api.rb [hunk 3]
```diff
@@ -0,0 +1,245 @@
...
+        elsif !options[:username].nil? && !options[:password].nil?
+          @ticket_granting_ticket = get_ticket_granting_ticket(options[:username], options[:password])
+        end
+      end
+
+      ##
+      # Gets the list of profiles. This may be used without credentials.
+      #
+      # Returns a list of profile names.
+      def get_profiles
```
Comment (jbradl11) [inline lib/util/vsac_api.rb]: Do we know how the profiles and programs are ordered when we get back the responses? If so we should document that in the comment above each of these functions.

File: lib/util/vsac_api.rb [hunk 4]
```diff
@@ -0,0 +1,245 @@
...
+      #
+      # Returns the JSON parsed response for program details.
+      def get_program_details(program = nil)
+        # if no program was provided use the one in the config or default in constant
+        if program == nil
+          program = @config.fetch(:program, DEFAULT_PROGRAM)
+        end
+
+        # parse json response and return it
+        return JSON.parse(RestClient.get("#{@config[:utility_url]}/program/#{URI.escape(program)}"))
```
Comment (jbradl11) [inline lib/util/vsac_api.rb]: If the `program` that gets passed in is invalid, what error response would this return? Should the api read the response and throw an error if it is invalid? This question also applies to the `get_releases_for_program` function below.

Comment (hossenlopp) [reply lib/util/vsac_api.rb]: Good catch. I'll make it handle that.

File: lib/util/vsac_api.rb [hunk 5]
```diff
@@ -0,0 +1,245 @@
...
+
+      ##
+      # Gets a valueset. This requires credentials.
+      #
+      def get_valueset(oid, options = {})
+        # base parameter oid is always needed
+        params = { id: oid }
+
+        # release parameter, should be used moving forward
+        if !options[:release].nil?
```
Comment (jbradl11) [inline lib/util/vsac_api.rb:202]: For consistency make this ` if options.has_key?(:release)`

File: lib/util/vsac_api.rb [hunk 6]
```diff
@@ -0,0 +1,245 @@
...
+      ##
+      # Gets the list of profiles. This may be used without credentials.
+      #
+      # Returns a list of profile names.
+      def get_profiles
+        profiles_response = RestClient.get("#{@config[:utility_url]}/profiles")
+        profiles = []
+
+        # parse xml response and get text content of each profile element
+        doc = Nokogiri::XML(profiles_response)
```
Comment (jbradl11) [inline lib/util/vsac_api.rb:126]: Looks like some other API calls return JSON responses, is it not possible to have all of the responses return as JSON or XML? If not...that's annoying.

Comment (hossenlopp) [reply lib/util/vsac_api.rb:126]: Nope. I attempted the Accept header. No luck.

File: lib/util/vsac_api.rb [hunk 7]
```diff
@@ -0,0 +1,245 @@
...
+
+      def get_ticket
+        # if there is no ticket granting ticket then we should raise an error
+        raise VSACNoCredentialsError.new unless @ticket_granting_ticket
+        # if the ticket granting ticket has expired, throw an error
+        raise VSACTicketExpiredError.new if Time.now > @ticket_granting_ticket[:expires]
+
+        # attempt to get a ticket
+        begin
+          ticket = RestClient.post("#{@config[:auth_url]}/Ticket/#{@ticket_granting_ticket[:ticket]}", service: "http://umlsks.nlm.nih.gov")
```
Comment (jbradl11) [inline lib/util/vsac_api.rb]: Should `http://umlsks.nlm.nih.gov` be set as a variable somewhere or in a config? If you don't think its necessary, at least add a comment explaining why its fine as a hardcoded string (example: "This web-url will never change")

Comment (hossenlopp) [reply lib/util/vsac_api.rb]: I'll move it to a constant with a comment why it will never change.

File: lib/util/vsac_api.rb (hunk 1 above)
Comment (hossenlopp) [reply lib/util/vsac_api.rb]: The all VSAC parameters are required at time of requesting an valueset. None of them can be specified in the config.

There is no default release constant or config option. The release parameter is used without a program parameter when getting a value set by release. I presume VSAC does not allow for releases of the same names in different programs.

File: lib/util/vsac_api.rb (hunk 3 above)
Comment (hossenlopp) [reply lib/util/vsac_api.rb]: They are ordered and we are preserving the order. I'll comment about that.

File: lib/util/vsac_api.rb (hunk 1 above)
Comment (hossenlopp) [reply lib/util/vsac_api.rb]: The `default_profile` in the `APP_CONFIG` is needed by the code that calls the VSAC API for if the front end passes down `include_draft` without a `profile`. I plan to remove this in the future and expect the front end to always pass down `profile` if it wants to `include_draft`.

File: lib/util/vsac_api.rb [hunk 8]
```diff
@@ -0,0 +1,264 @@
...
+      # Returns the JSON parsed response for program details.
+      def get_program_details(program = nil)
+        # if no program was provided use the one in the config or default in constant
+        if program == nil
+          program = @config.fetch(:program, DEFAULT_PROGRAM)
+        end
+
+        begin
+          # parse json response and return it
+          return JSON.parse(RestClient.get("#{@config[:utility_url]}/program/#{URI.escape(program)}"))
```
Comment (lukoktonos) [inline lib/util/vsac_api.rb]: Rubocop warning:
```
lib/util/vsac_api.rb:164:80: W: Lint/UriEscapeUnescape: URI.escape method is obsolete and should not be used. Instead, use CGI.escape, URI.encode_www_form orURI.encode_www_form_component depending on your specific use case.
```

File: lib/util/vsac_api.rb [hunk 9]
```diff
@@ -0,0 +1,264 @@
...
+        if !options[:version].nil?
+          params[:version] = options[:version]
+        end
+
+        # get a new service ticket
+        params[:ticket] = get_ticket
+
+        # run request
+        begin
+          value_set_response = RestClient.get("#{@config[:content_url]}/RetrieveMultipleValueSets", params: params)
```
Comment (lukoktonos) [inline lib/util/vsac_api.rb]: Rubocop warning:
```
lib/util/vsac_api.rb:220:11: W: Lint/UselessAssignment: Useless assignment to variable - value_set_response.
          value_set_response = RestClient.get("#{@config[:content_url]}/RetrieveMultipleValueSets", params: params)
          ^^^^^^^^^^^^^^^^^^
```

File: lib/util/vsac_api.rb [hunk 10]
```diff
@@ -0,0 +1,264 @@
...
+        end
+
+        return profiles
+      end
+
+      ##
+      # Gets the list of programs. This may be used without credentials.
+      #
+      # Returns a list of program names. These are kept in the order that VSAC provides them in.
+      def get_programs
```
Comment (lukoktonos) [inline lib/util/vsac_api.rb]: Recommend re-naming this to `get_program_names`

File: lib/util/vsac_api.rb [hunk 11]
```diff
@@ -0,0 +1,264 @@
...
+        elsif !options[:username].nil? && !options[:password].nil?
+          @ticket_granting_ticket = get_ticket_granting_ticket(options[:username], options[:password])
+        end
+      end
+
+      ##
+      # Gets the list of profiles. This may be used without credentials.
+      #
+      # Returns a list of profile names. These are kept in the order that VSAC provides them in.
+      def get_profiles
```
Comment (lukoktonos) [inline lib/util/vsac_api.rb]: recommend renaming this to `get_profile_names`

File: lib/util/vsac_api.rb [hunk 12]
```diff
@@ -0,0 +1,264 @@
...
+        rescue RestClient::ResourceNotFound
+          raise VSACProgramNotFoundError.new(program)
+        end
+      end
+
+      ##
+      # Gets the releases for a program. This may be used without credentials.
+      #
+      # Returns a list of releases in a program. These are kept in the order that VSAC provides them in.
+      def get_releases_for_program(program = nil)
```
Comment (lukoktonos) [inline lib/util/vsac_api.rb]: recommend renaming to `get_program_releases` to match above functions

File: lib/util/vsac_api.rb [hunk 13]
```diff
@@ -0,0 +1,264 @@
...
+        # release parameter, should be used moving forward
+        if !options[:release].nil?
+          params[:release] = options[:release]
+        end
+
+        # profile parameter, may be needed for getting draft value sets
+        if !options[:profile].nil?
+          params[:profile] = options[:profile]
+          if !options[:include_draft].nil?
+            params[:includeDraft] = if !!options[:include_draft] then 'yes' else 'no' end
```
Comment (lukoktonos) [inline lib/util/vsac_api.rb:210]: Hadn't seen the `!!` before in ruby--interesting but a bit confusing if not familiar with it.

File: lib/util/vsac_api.rb (hunk 12 above)
Comment (hossenlopp) [reply lib/util/vsac_api.rb]: Shouldn't it be `get_program_release_names` because it will return the list of release names for a given program.

Comment (lukoktonos) [reply lib/util/vsac_api.rb]: Yes that would be even better--and updating the comment to include names as well would make them match.

File: lib/util/vsac_api.rb (hunk 13 above)
Comment (hossenlopp) [reply lib/util/vsac_api.rb:210]: It's a way that I've seen people cast things to boolean. This may actually be a bad thing to do...

File: lib/util/vsac_api.rb (hunk 5 above)
Comment (hossenlopp) [reply lib/util/vsac_api.rb:202]: Per slack discussion we decided on `!options[:symbol].nil?` across this class.

File: lib/util/vsac_api.rb (hunk 13 above)
Comment (holmesie) [reply lib/util/vsac_api.rb:210]: `!!` is a thing in most languages, and while I don't think it is necessary here (`nil` would result in `no` just as well as `false` would), I think its use tends to make things more explicit.

Comment (hossenlopp) [reply lib/util/vsac_api.rb:210]: So, should i just completely remove `!!`?

Comment (holmesie) [reply lib/util/vsac_api.rb:210]: I'd say leave it

**Review by lukoktonos — APPROVED** [review]
Looks good. CI failing due to gem versions that will be fixed when `bonnie-v2.1` is updated to master.

**Review by jbradl11 — APPROVED** [review]
I am happy with this refactor. Good job!

**Comment by edeyoung** [comment]
failing due to vulnerability that will be addressed prior to any publication
